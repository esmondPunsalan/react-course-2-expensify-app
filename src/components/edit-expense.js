import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './expense-form';
import RemoveModal from './remove-modal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            removeRequest: false
        };
    }
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    }
    onRemoveRequest = () => {
        this.setState(() => ({ removeRequest: true }));
    }
    closeRemoveModal = (e) => {
        const result = e.target.value == 1 ? true : false;
        this.setState(() => ({ removeRequest: false }));

        if(result) {
            this.props.startRemoveExpense({ id: this.props.expense.id });
            this.props.history.push("/");
        }
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                        expense={this.props.expense}
                        onSubmit={ this.onSubmit }/>
                    <button className="button button--secondary" onClick={this.onRemoveRequest}>Remove</button>
                </div>
                <RemoveModal removeRequest={this.state.removeRequest} closeRemoveModal={this.closeRemoveModal} />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);