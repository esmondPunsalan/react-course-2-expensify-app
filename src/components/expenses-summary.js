import React from "react";
import numeral from "numeral"
import getExpensesTotal from "../selectors/expenses-total";
import selectExpenses from '../selectors/expenses';
import { connect } from "react-redux";
import { Link } from "react-router-dom";

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
    const expenseWord = expensesCount == 1 ? "expense" : "expenses";
    const formatedTotal = numeral(expensesTotal / 100).format("$0,0.00");

    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span>{expensesCount}</span> {expenseWord} totalling <span>{formatedTotal}</span></h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create" >+ Expense</Link>
                </div>
            </div>
        </div>
    )
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    
    return { 
        expensesCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    };
}

export default connect(mapStateToProps)(ExpensesSummary);