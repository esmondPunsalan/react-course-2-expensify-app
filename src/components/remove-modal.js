import React from "react";
import Modal from "react-modal";

export default class RemoveModal extends React.Component {

    componentWillMount() {
        Modal.setAppElement('body');
    }

    render () {
        return (
            <Modal
                isOpen={!!this.props.removeRequest}
                onRequestClose={this.props.closeRemoveModal}
                className="modal"
            >
                <div>Are you sure you want to remove this expense?</div>
                <div>
                    <button className="button modal__button" onClick={this.props.closeRemoveModal} value="1">Yes</button>
                    <button className="button modal__button" onClick={this.props.closeRemoveModal} value="0">No</button>
                </div>
            </Modal>
        )
    }
}