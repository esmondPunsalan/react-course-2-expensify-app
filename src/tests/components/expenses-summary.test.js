import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/expenses-summary";

test("should render expense summary component correctly with empty array", () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={0} expensesTotal={0}/>);
    expect(wrapper).toMatchSnapshot();
});

test("should render expense summary component correctly with data", () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={3} expensesTotal={459875}/>);
    expect(wrapper).toMatchSnapshot();
});