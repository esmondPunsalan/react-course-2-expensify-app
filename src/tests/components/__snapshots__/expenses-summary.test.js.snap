// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary component correctly with data 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        $4,598.75
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        + Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render expense summary component correctly with empty array 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totalling 
      <span>
        $0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        + Expense
      </Link>
    </div>
  </div>
</div>
`;
